src/coiio/imageio.cpp

```diff
@@ -0,0 +1,201 @@
+}
+
+bool
+ImageSpec_getattribute(const ImageSpec* is, const char* name, TypeDesc type,
+                       void* value, bool casesensitive)
+{
+    return is->getattribute(name, pun<OIIO::TypeDesc>(type), value,
+                            casesensitive);
+}
+
+// FIXME: add Filesystem::IOProxy
+ImageInput*
+ImageInput_open(const char* filename, const ImageSpec* config, void* _ioproxy)
+{
+    auto ii = ImageInput::open(filename, config, nullptr);
+    return ii.release();
+}
+
+bool
+ImageInput_close(ImageInput* ii)
+{
+    return ii->close();
+}
+
+void
+ImageInput_delete(ImageInput* ii)
+{
+    delete ii;
+}
+
+const ImageSpec*
+ImageInput_spec(ImageInput* ii)
+{
+    return &ii->spec();
+}
+
+bool
+ImageInput_read_image(ImageInput* ii, int subimage, int miplevel, int chbegin,
+                      int chend, TypeDesc format, void* data, stride_t xstride,
+                      stride_t ystride, stride_t zstride,
+                      ProgressCallback progress_callback,
+                      void* progress_callback_data)
+{
+    return ii->read_image(subimage, miplevel, chbegin, chend,
+                          pun<OIIO::TypeDesc>(format), data, xstride, ystride,
+                          zstride, progress_callback, progress_callback_data);
+}
+
+OIIO_API bool
+ImageInput_has_error(const ImageInput* ii)
+{
+    return ii->has_error();
+}
+
+const char*
+ImageInput_geterror(const ImageInput* ii)
+{
+    // If we have multiple ImageInputs that we want to get errors for, we're
+    // still passing through this one function. We have to cache the string
+    // in order to return a char*, so it needs to be thread_local to avoid issues.
+    // We force clear here so that multiple calls from different ImageInputs
+    // don't get each others' errors.
+    thread_local std::string errorstring;
+    errorstring = ii->geterror(true);
+    return errorstring.c_str();
+}
+
+ImageOutput*
+ImageOutput_create(const char* filename, void* ioproxy,
+                   const char* plugin_search_path)
+{
+    return ImageOutput::create(filename, nullptr, plugin_search_path).release();
+}
+
+void
+ImageOutput_delete(ImageOutput* io)
+{
+    delete io;
+}
+
+bool
+ImageOutput_open(ImageOutput* io, const char* name, const ImageSpec* newspec,
+                 int mode)
+{
+    return io->open(std::string(name), *newspec,
+                    (OIIO::ImageOutput::OpenMode)mode);
+}
+
+OIIO_API bool
+ImageOutput_has_error(const ImageOutput* io)
+{
+    return io->has_error();
+}
+
+const char*
+ImageOutput_geterror(const ImageOutput* io)
+{
+    // If we have multiple ImageOutputs that we want to get errors for, we're
+    // still passing through this one function. We have to cache the string
+    // in order to return a char*, so it needs to be thread_local to avoid issues.
+    // We force clear here so that multiple calls from different ImageOutputs
+    // don't get each others' errors.
+    thread_local std::string errorstring;
```

Comment:
I'm 99% sure thread_local implies static, such that `static thread_local std::string` and `thread_local std::string` are equivalent. That said, it would't therefore hurt to add `static` so that others (and future me) don't have the same reaction.

Reply:
Oh, you could be right. But in any case, the thread safety, where it's needed, should all be handled on the C++ side. I don't think you should ever need to impose additional thread safety on the C side.

ImageOutput is not thread safe at all, it doesn't mean anything useful for threads to simultaneously output to one file. ImageInput, though, should be thread safe (makes perfect sense for multiple threads to read simultaneously from the same open file), so if there's someplace there where we slipped up on the C++ side, we should address that separately.